### ISSUE TITLE
[keyname] behaves differently when patch is modified (Mac Big Sur 11.6.6)

### ISSUE THREAD

**Issue by ArturJD96** [pure-data/pure-data#1845]
When a keyboard key is continuously pressed, left outlet of [keyname] (reporting <KeyPress> and <KeyRelease> events as 1 and 0) behaves differently when the patch has been just opened and after edition (e.g. manually creating objects)

**Behaviour 1** (_undesirable_) – patch has been just opened
* key is pressed
* left outlet of [keyname]: 1 0 1 0 1 0 1 0 1 0 ...

**Behaviour 2** (_correct_) – patch has been manually modified (e.g. object created) after opening
* key is pressed
* left outlet of [keyname]: 1 1 1 1 1 1 1 1 1 ...

Note: the patch must be manually edited after opening to get the right behaviour; dynamically creating objects with pd-messages and [loadbang] in order to force behaviour 2 doesn't work.

Possibly related to: https://github.com/pure-data/pure-data/issues/213
[keynameIssue.txt](https://github.com/pure-data/pure-data/files/10146722/keynameIssue.txt)

(I attach pd testing patch as a .txt file).

**Comment by jamshark70** [comment]
FWIW I tested in Linux and I get behavior 1 in all cases (never got behavior 2).

So the discrepancy may be OS-specific. This is often -- *often* -- the case for UI issues. @ArturJD96 which OS are you using?

**Comment by sebshader** [comment]
@jamshark70 in the title it says MacOs 11.6.6.
I assume tcl/tk 8.6.10? Might be worth checking if it's fixed in 8.6.12..
I don't remember case 1 happening under mojave (10.14) on either. (I think it was always 1 1 1 1 etc.)

**Comment by ArturJD96** [comment]
It turns out that I have the 8.5.9 version of tcl/tk and this might be the problem indeed. I'll try to upgrade and inform about the results.

**Comment by sebshader** [comment]
@ArturJD96 look at the pdwindow with the most verbose log level when pd starts to see the tcl/tk version
Pd ships with a different version than the system ships with

**Comment by ArturJD96** [comment]
It seems that Apple ceases to support Tk altogether. When typing "tclsh" in the terminal (by means of which we enter the tcl), the following warning appears:

WARNING: This version of tcl is included in macOS for compatibility with legacy software. 
In future versions of macOS the tcl runtime will not be available by 
default, and may require you to install an additional package.

Also, it turns out that updating tk on Mac is not straightforward anymore: https://discussions.apple.com/thread/250803178

**Comment by ArturJD96** [comment]
@sebshader, thank you for suggestion, it turns out that I have Tk 8.6.12 version working with PD.

**Comment by FlachyJoe** [comment]
Hi,
Case 1 for me
```
XUbuntu 22.04
Pd version 0.52.1
tk version 8.6.11
```
![Capture d’écran_2022-12-04_21-00-49](https://user-images.githubusercontent.com/10371513/205512700-3338f85c-2956-4a09-9f4a-a12ccc39299f.png)

**Comment by sebshader** [comment]
@FlachyJoe again, this difference in behavior between MacOS and linux has always been the case (since the key-repeat fix in tcl/tk on MacOS anyways)

Again, iirc '1 1 1 1 0' is the expected behavior on MacOs and '1 0 1 0 1 0 1 0 ' is the expected behavior on linux

**Comment by sebshader** [comment]
@ArturJD96 I don't know if this has to be said, but if you set key repeats to 'off' in the system preferences you shouldn't get any repeats when holding a key down, after restarting pd

**Comment by sebshader** [comment]
also could someone see what the behavior of the test patch from the linked issue is when using an 8.6.12 Wish.app? (and if it isn't the same behavior, if it occurs when used to package pd?)
```
package require Tk

bind . <KeyPress>   {puts "press   %K"}
bind . <KeyRelease> {puts "release %K"}
```
@danomatika?

**Comment by ArturJD96** [comment]
@sebshader:
Yes, I know, but I don't want to rely on that setting – handling it in pd would be much safer IMO. What makes it buggy is indeed the fact that the same object has two discrepant behaviour. It is good to know about the influence of the operating system on this object.

@FlachyJoe: that's a valuable insight! might be related.

**Comment by sebshader** [comment]
@ArturJD96 the behavior that @FlachyJoe describes is what is supposed to happen on linux.. because it is generally the responsibility of the OS to deliver key events

**Comment by ArturJD96** [comment]
@sebshader: when window is not focused, [keyname] does not output at all.

**Comment by danomatika** [comment]
Better

> also could someone else reproduce & then see what the behavior of the test .tcl file from the linked issue is when using a 8.6.12 Wish.app? (and if it isn't the same behavior, if it occurs when used to package pd?)
> 
> ```
> package require Tk
> 
> bind . <KeyPress>   {puts "press   %K"}
> bind . <KeyRelease> {puts "release %K"}
> ```
> 
> would anyone possibly feel like testing/reproducing? @danomatika @porres?

What am I testing? If there are distinct press/release events? There are with the System key repeat enabled and disabled.

Expected behavior for Pd on macOS should be:
* key repeat enabled: 1 0 1 0 1 0... at some point .... 0 (generated on/off)
* key repeat disabled: 1 .... at some point ...0 (one pair)

This appears to be correct in the simple Tcl script, so I imagine it has more to do with the key press/release handling in `tcl/pd_bindings.tcl`: https://github.com/pure-data/pure-data/blob/master/tcl/pd_bindings.tcl#L390 Best to ask @umlaeute

Also, please keep in mind that fixing key event handling has been a major PITA these last years on macOS. It may take some extra sleuthing to convince everyone to spend more time on it right now...

**Comment by danomatika** [comment]
> @FlachyJoe Again, iirc '1 1 1 1 0' is the expected behavior on MacOs and '1 0 1 0 1 0 1 0 ' is the expected behavior on linux (since the key-repeat patch for tcl/tk on MacOS anyways)
> 
> this aligns with the behavior of utilities that capture key presses in the terminals on these systems

I think this is incorrect. The repeated 1s were the result of incomplete keyboard handling in TK's move form Carbon to Cocoa, which we've had to rectify via custom patching in the past. Ideally, repeats should generate distance pairs in Pd, I believe,

**Comment by sebshader** [comment]
oh maybe you're right. if so sorry for aggressively misremembering.

**Comment by sebshader** [comment]
@danomatika I was just wondering if this behavior happens on other (older) MacOS versions because if not it's probably a tcl/tk bug

**Comment by danomatika** [comment]
> @danomatika I was just wondering if this behavior happens on other (older) MacOS versions

Try an old version of Pd, maybe 0.47 or earlier on an older system. It should still be using Tk 8.4 or so. They switched to the Cocoa APIs around 8.5.10 I think.
